spmd/tensor/device_mesh.py

```diff
@@ -396,14 +396,35 @@ def all_gather(
         Returns:
             A :class:`torch.Tensor` object
         """
+        my_coordinate = self.get_coordinate_on_dim(mesh_dim)
+        # TODO: what should happen if rank is not in the mesh?
+        assert (
+            my_coordinate is not None
+        ), "Rank if not part of mesh"  # TODO: figure out behavior here
+
         num_chunks = self.size(mesh_dim)
-        _, rem = divmod(output_shape[tensor_dim], num_chunks)
-        assert rem == 0, "output_shape must be divisible by num_chunks"
-        input_tensor = tensor.contiguous()
-        gathered_list = [
-            CommTensor(torch.empty_like(input_tensor))
-            for _ in range(num_chunks)
-        ]
+        quot, rem = divmod(output_shape[tensor_dim], num_chunks)
+        gathered_list = []
+        for _ in range(num_chunks):
+            recv_shape = list(output_shape)
+            # create recv tensor with padded shape
+            recv_shape[tensor_dim] = quot + (1 if rem > 0 else 0)
```

Comment:
Should we reuse local_shape() or somehow consolidate here our formula for shapes?

Reply:
good point! Yeah we should consolidate local shape calculation, the tricky part on this is that it involves padded shape, i think we can abstract this local shape + padded logic out and make a common calculation, i'll do this in my next PR about the uneven sharding